File: src/main/java/architecture/ego_equipment/datagen/DatagenSoundDefinitionsProvider.java
package architecture.ego_equipment.datagen;

import architecture.ego_equipment.core.EGOEquipment;
import architecture.ego_equipment.init.EGOEquipmentSoundEvents;
import net.minecraft.data.PackOutput;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.sounds.SoundEvent;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.common.data.SoundDefinition;
import net.neoforged.neoforge.common.data.SoundDefinitionsProvider;
import org.jetbrains.annotations.NotNull;

public class DatagenSoundDefinitionsProvider extends SoundDefinitionsProvider {
	public DatagenSoundDefinitionsProvider(PackOutput output, ExistingFileHelper existingFileHelper) {
		super(output, EGOEquipment.ID, existingFileHelper);
	}

	public static @NotNull String getSubtitle(SoundEvent soundEvent) {
		return getSubtitle(soundEvent.getLocation());
	}

	public static @NotNull String getSubtitle(ResourceLocation location) {
		return "sound." + location.toLanguageKey();
	}

	@Override
	public void registerSounds() {
		add(EGOEquipmentSoundEvents.SOLEMN_LAMENT_WEAPON_ATTACK_BLACK.value(), 0.5f, 0.5f, 1, 8);
		add(EGOEquipmentSoundEvents.SOLEMN_LAMENT_WEAPON_ATTACK_WHITE.value(), 0.5f, 0.5f, 1, 8);
		add(EGOEquipmentSoundEvents.SOLEMN_LAMENT_WEAPON_STONGATTACK_BLACK.value(), 0.5f, 0.5f, 1, 8);
		add(EGOEquipmentSoundEvents.SOLEMN_LAMENT_WEAPON_STONGATTACK_WHITE.value(), 0.5f, 0.5f, 1, 8);
	}

	private void add(SoundEvent soundEvent, float volume, float pitch, int weight, int attenuationDistance) {
		add(soundEvent, 1, volume, pitch, weight, attenuationDistance);
	}

	private void add(SoundEvent soundEvent, int amount, float volume, float pitch, int weight, int attenuationDistance) {
		ResourceLocation location = soundEvent.getLocation();
		assert amount > 0 : "sund : " + location + " amount must be greater than 0";
		SoundDefinition definition = SoundDefinition.definition();
		for (int i = 0; i < amount; i++) {
			definition.with(SoundDefinition.Sound.sound(location.withSuffix(i > 0 ? String.valueOf(i) : ""), SoundDefinition.SoundType.SOUND)
				.volume(volume)
				.pitch(pitch)
				.weight(weight)
				.attenuationDistance(attenuationDistance)
			);
		}
		add(soundEvent, definition
			.subtitle(getSubtitle(location))
			.replace(true)
		);
	}
}

File: src/main/kotlin/architecture/ego_equipment/datagen/DatagenSoundDefinitionsProvider.kt
package architecture.ego_equipment.datagen

import architecture.ego_equipment.init.EgoEquipSoundEvents
import architecture.ego_equipment.util.EgoEquipUtil
import architecture.goldenboughs_lib.datagen.BasicSoundDefinitionsProvider
import net.minecraft.data.PackOutput
import net.minecraft.resources.ResourceLocation
import net.minecraft.sounds.SoundEvent
import net.neoforged.neoforge.common.data.ExistingFileHelper

class DatagenSoundDefinitionsProvider(output: PackOutput, existingFileHelper: ExistingFileHelper) :
	BasicSoundDefinitionsProvider(output, EgoEquipUtil.ID, existingFileHelper) {

	companion object {
		@JvmStatic
		fun getSubtitle(soundEvent: SoundEvent): String = getSubtitle(soundEvent.location)

		@JvmStatic
		fun getSubtitle(location: ResourceLocation): String = "sound.${location.toLanguageKey()}"
	}

	override fun registerSounds() {
		add(EgoEquipSoundEvents.SOLEMN_LAMENT_WEAPON_ATTACK_BLACK.value(), 0.5f, 0.5f, 1, 8)
		add(EgoEquipSoundEvents.SOLEMN_LAMENT_WEAPON_ATTACK_WHITE.value(), 0.5f, 0.5f, 1, 8)
		add(EgoEquipSoundEvents.SOLEMN_LAMENT_WEAPON_STONGATTACK_BLACK.value(), 0.5f, 0.5f, 1, 8)
		add(EgoEquipSoundEvents.SOLEMN_LAMENT_WEAPON_STONGATTACK_WHITE.value(), 0.5f, 0.5f, 1, 8)
	}
}
